### PR TITLE
Support FS_CASEFOLD_FL (F) file attribute on Linux

### PR DESCRIPTION
**Pull request by ryao**
### Motivation and Context
I got a steam deck recently and examined how it formatted the microsd card that I gave it. It is formatted as ext4 with an interesting feature called casefold, which was introduced in Linux 5.2. That feature allows empty directories to be marked as case insensitive on what is an otherwise case sensitive filesystem. Userland will then have case insensitive lookups after the directory is populated and software such as wine that typically does expensive emulation of case insensitive directory lookups can instead check the file attribute on directories to discover opportunities to offload the lookups to the filesystem, which improves performance.

https://github.com/wine-mirror/wine/blob/cb8937801d42417c2b98cba15366ccc154c1cc40/dlls/ntdll/unix/file.c#L1166
https://github.com/ValveSoftware/Proton/blob/4221d9ef07cc38209ff93dbbbca9473581a38255/proton#L390

Anyway, I felt inspired upon seeing ext4's casefold extension to take a look at whether we could support additional Linux file attributes in ZFS and it turns out that we can. `FS_CASEFOLD_FL`, `FS_INLINE_DATA_FL` and `FS_ENCRYPT_FL` all can be implemented as virtual file attributes. While `FS_CASEFOLD_FL` is not implemented as a virtual file attribute in other filesystems, implementing it that way is compatible with both how current and future userland software is expected to interact with it. The latter two are actually virtual file attributes on other filesystems, but they are more difficult to implement since implementing them appears to require passing metadata from the DMU to the ZPL. Thus, I have only implemented `FS_CASEFOLD_FL` in this patch.

### Description
This is a tiny change to the ZPL to implement the `FS_CASEFOLD_FL` file attribute. It only affects case insensitive filesystems. case sensitive filesystems and mixed filesystems are unaffected.

The semantics of this implementation on case
insensitive filesystems are as follows:

 * Directories always report `FS_CASEFOLD_FL` through the `FS_IOC_GETFLAGS`
   ioctl while non-directories never report it.

 * The `FS_IOC_SETFLAGS` ioctl accepts the `FS_CASEFOLD_FL` file attribute
   on directories, but rejects it on non-directories by returning an
   `ENOTDIR`.

 * The `FS_IOC_SETFLAGS` ioctl silently ignores attempts to unset the
   `FS_CASEFOLD_FL` file attribute. This theoretically maintains
   compatibility with userland tools that copy directories from case
   sensitive (ZFS) filesystems and do not read the file attribute
   bitmask before setting it. This is also consistent with the behavior
   of other Linux filesystems when given file attribute bitmasks missing
   file attributes that userland may not unset.

This is compatible both with how proton is programmed to set the
`FS_CASEFOLD_FL` file attribute and with how wine reads the
`FS_CASEFOLD_FL` file attribute.  It differs from the ext4 semantics in
the following ways:

 * ext4 does not create directories with this file attribute set on
   casefold ext4 filesystems. All directories on case insensitive ZFS
   filesystems have this file attribute implicitly set.

 * ext4 will return `ENOTEMPTY` when userland attempts to set this bit
   on non-empty directories on casefold ext4 filesystems. Since the file
   attribute is always set on directories on case insensitive ZFS
   filesystems, it does not make sense to return `ENOTEMPTY`.
   Furthermore, having this bit always implicitly set interacts with the
   design of the Linux interface for setting file attributes in such a
   way that we never encounter the case to be able to return `ENOTEMPTY`
   even if we wanted to return it.

 * ext4 allows the bit to be unset on empty directories. We silently
   accept the unset operation on case insensitive filesystems, but we do
   not "unset" the bit.

There is a theoretical edge case that occurs if userland attempts to
copy a directory from a case insensitive filesystem to a case sensitive
filesystem while preserving file attributes. In that situation, a naive
attempt to set the original bitmask would be rejected, which could cause
the loss of file attributes unless userland falls back to setting 1 bit
at a time to preserve as many bits as possible. Any directories with the
immutable bit set would need to have more intelligent treatment (such as
setting it after creating the files inside the directory), so expecting
userland to be smart is reasonable and we do not have to worry about
this edge case.

The semantics of this implementation are sufficiently compatible with
both current userland software and any foreseeable future userland
software that I feel comfortable implementing this file attribute in ZFS
on Linux.

Most of the above description is copied from the commit message. There is a little more written there, although nearly everything except a remark about F2FS also supporting this file attribute has been rewritten here in some form. There are also comments inline in the code mentioning things that I felt were non-obvious.

### How Has This Been Tested?

I have only done a local build test of the Linux ZFS kernel modules from git master with this patch applied against Linux 5.16.14-gentoo-x86_64.

I have done no functional testing. Instead, I have written test cases and included them with the changes. The test cases are expected to enable the buildbot to validate that this change is working as expected. The test cases have the limitation that they only test for errors and do not check if the error codes are what are expected.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by ryao — COMMENTED** [review]
This is my first time writing test cases, so I used an existing test case as a template and another as a guide. I then retained the copyright notices from both of them. I also copy and pasted a copyright notice that I used in the past for my work (it is a LLC I own) and updated the year for it.

File: tests/zfs-tests/tests/functional/chattr/chattr_004_neg.ksh [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+#	Verify FS_CASEFOLD_FL is unsupported on case sensitive filesystems.
+#
+# STRATEGY:
+#	1. Create a filesystem with casesensitivity=insensitive.
```
Comment (gmelikov) [inline tests/zfs-tests/tests/functional/chattr/chattr_004_neg.ksh:41]:
```suggestion
#	1. Create a filesystem with casesensitivity=sensitive.
```

Comment (ryao) [reply tests/zfs-tests/tests/functional/chattr/chattr_004_neg.ksh:41]: Nice find. Thanks. I just amended my commit locally. I will repush after posting this comment.

**Comment by ryao** [comment]
```
make: *** [Makefile:13751: commitcheck] Error 1
make: *** Waiting for unfinished jobs....
error: commit message body contains line over 72 characters
Error: Process completed with exit code 2.
```

The github urls I included in the commit message are huge. I am not sure how to include links to relevant code in the commit message without triggering this. I see recent cases in the commit history where commits were made that violated the 72 character per line limit. I imagine this could be another of them.

**Comment by ryao** [comment]
```
/sbin/ztest -G -VVVVV -K draid -m 0 -r 18 -D 10 -S 3 -R 2 -v 0 -a 12 -C special=random -s 512m -f /var/tmp/zloop-run -T 120 -P 60
*** ztest crash found - moving logs to /var/tmp/zloop/zloop-220822-205940

warning: Can't open file /tmp/ztest.data.vDDUe1 (deleted) during file-backed mapping note processing
*** core @ /var/tmp/zloop/zloop-220822-205940/core:
```

I did not touch any code that ztest would exercise, so I am going to assume this is a pre-existing bug.

**Comment by ryao** [comment]
Upon another look at the wine source code, it turns out that wine *does* benefit from this file attribute, although why proton did not mark the directories as being case insensitive on my microsd card is a minor mystery that I do not plan to pursue. I have updated the description by deleting the remark about using this appearing to be a planned future enhancement.

Also, I had leaned on the buildbot for testing since I do not currently have my original development environment available at the moment. I had assumed that the korn shell test cases were simple, so I wrote some that turned out to be subtly wrong. That is not a surprise given that I did no local testing since I had decided to try outsourcing testing to the buildbot. What was a surprise is just how much I found myself to enjoy relying on the buildbot to run the test cases for me. In the past, I would manually check a significant subset of the cases that the buildbot could have checked for me and I found that process to be mentally tiring. This week, outsourcing that busywork to the buildbot, even with mistakes in the initial test cases, freed my mind to focus on more interesting things, so I have none of the mental fatigue that I used to have from checking my work. I wish I had begun to write ZTS test cases years ago. :)

**Comment by gmelikov** [comment]
@ryao offtopic, but if you enable github actions in your fork - you can even outsource tests there without any PR too, just in case :) And you can easily patch https://github.com/openzfs/zfs/blob/master/.github/workflows/zfs-tests-sanity.yml#L59 to run exactly what you want.

**Comment by ryao** [comment]
> @ryao offtopic, but if you enable github actions in your fork - you can even outsource tests there without any PR too, just in case :) And you can easily patch https://github.com/openzfs/zfs/blob/master/.github/workflows/zfs-tests-sanity.yml#L59 to run exactly what you want.

Thanks for the tip. I will be sure to try that for my next pull request that can benefit from it. :)

**Comment by ryao** [comment]
@behlendorf At this point, the test failures are pre-existing issues (and one false positive since the commit style check cannot exclude URLs from the 72 character limit check). Unless anyone has any style criticisms that the automated style check missed or wants revisions to the tests, this should be ready to merge.

File: module/os/linux/zfs/zpl_file.c [hunk 2]
```diff
@@ -963,6 +981,25 @@ __zpl_ioctl_setflags(struct inode *ip, uint32_t ioctl_flags, xvattr_t *xva)
 	uint64_t zfs_flags = ITOZ(ip)->z_pflags;
 	xoptattr_t *xoap;
 
+	zfsvfs_t *zfsvfs = ip->i_sb->s_fs_info;
```
Comment (behlendorf) [inline module/os/linux/zfs/zpl_file.c:984]:
```suggestion
	zfsvfs_t *zfsvfs = ITOZSB(ip);
```

File: module/os/linux/zfs/zpl_file.c [hunk 3]
```diff
@@ -906,15 +906,33 @@ zpl_ioctl_getversion(struct file *filp, void __user *arg)
...
+	 * On case insensitive filesystems, we report the ZFS_CASEFOLD_FL
+	 * attribute on directories so that userland software is informed that
+	 * directory lookups are case insensitive. Userland software, such as
+	 * wine, that emulates case insensitive directory lookups on Linux
+	 * systems that typically have case sensitive directory lookups can use
+	 * this as a hint to learn that it can accelerate itself by offloading
+	 * the lookups to the filesystem.
+	 */
+	if (S_ISDIR(ITOZ(ip)->z_mode)) {
+		zfsvfs_t *zfsvfs = ip->i_sb->s_fs_info;
```
Comment (behlendorf) [inline module/os/linux/zfs/zpl_file.c:930]:
```suggestion
		zfsvfs_t *zfsvfs = ITOZSB(ip);
```

**Review by behlendorf — COMMENTED** [review]
This is neat, I hadn't noticed the addition of `FS_CASEFOLD_FL` to the kernel.  This looks good to me, just a few small nits.

File: include/os/linux/zfs/sys/zpl.h [hunk 4]
```diff
@@ -197,4 +197,6 @@ zpl_dir_emit_dots(struct file *file, zpl_dir_context_t *ctx)
 #define	zpl_setattr_prepare(ns, dentry, ia)	setattr_prepare(dentry, ia)
 #endif
 
+#define	ZFS_CASEFOLD_FL 0x40000000
```
[the remark this answers is not in the document]
Comment (ryao) [reply include/os/linux/zfs/sys/zpl.h:203]: Alright.

File: module/os/linux/zfs/zpl_file.c (hunk 2 above)
Comment (ryao) [reply module/os/linux/zfs/zpl_file.c]: I like this better. :)

File: module/os/linux/zfs/zpl_file.c (hunk 3 above)
Comment (ryao) [reply module/os/linux/zfs/zpl_file.c]: Ditto

**Comment by ryao** [comment]
@behlendorf I rebased on the latest master, made those small changes and repushed. :)

**Comment by ryao** [comment]
It occurs to me that Samba should be able to benefit from this too. A quick search of the Samba 4.16.2 suggests that it currently does not take advantage of it.

I do not plan to implement it in Samba right now, but if any Samba users out there are interested in making Samba directory lookups faster, modifying it to do what Wine does on Linux where it skips its case insensitive emulation following a missed lookup on directories with this file attribute would be a good idea. Interestingly, only directory lookup misses are accelerated. Directory lookup hits would not be accelerated since the fallback to emulate case insensitive directory lookups is not needed on hits.

**Review by behlendorf — APPROVED** [review]

**Review by gmelikov — APPROVED** [review]

**Comment by ryao** [comment]
In the interest of full disclosure, I have realized that this and ext4's implementation has a few more semantic differences involving unicode normalization and case sensitivity:

https://www.collabora.com/news-and-blog/blog/2020/08/27/using-the-linux-kernel-case-insensitive-feature-in-ext4/

The ext4 implementation does unicode normalization (formD). If a non-utf8 name is provided, ext4 will accept it, but case insensitivity would not work on it. In this patch, we do not require unicode normalization and thus accept non-utf8 names. I am not sure what happens to our case insensitivity in that case. If a user turns on unicode normalization at filesystem creation time, we instead reject non-utf8 names. ext4 optionally will reject non-utf8 names if its `strict` encoding flag is set.

In addition, the ext4 implementation will normalize `floß` to `floss`, which we currently do not do with unicode normalization and case insensitivity. This difference appears to occur because `casefold` is a reference to a python function that tries to more be a "more aggressive" version of `tolower()` to make case insensitive matches occur more frequently:

https://www.programiz.com/python-programming/methods/string/casefold

The only example I could find of it being "more aggressive" than `tolower()` is the conversion of `ß` to `ss`. A comment on that blog post mentions that normalizing `ß` to `ss` is actually wrong because `in Massen` and `in Maßen`, have opposite meanings.

Getting the exact unicode normalization semantics of ext4 would be a pain since:

* We would want to implement `casesensitivity=casefold` to convert `ß` to `ss` that also implies normalization=formD unless a different one is selected.
* We would want to allow `utf8only=off` on filesystems with case insensitivity, which I think would be a mistake.

That said, I believe that adopting strict adherence to the ext4 behavior would be wrong in the context of the userland consumers. The only userland software that uses this (and is likely to use this) would be things trying to emulate Windows directory lookups (e.g. wine, darling and samba) and while Windows directory lookups are case insensitive, they do not do unicode normalization (and thus consumers cannot expect `ß` to be equal to `ss`):

https://webtide.com/a-story-about-unix-unicode-java-filesystems-internationalization-and-normalization/

Therefore, tying case insensitivity to unicode normalization and implementing the more rigorous case sensitivity that treats `ß` to be equal to `ss` is detrimental to compatibility. At the same time, unicode normalization's applicability to Windows software is so rare that this likely does not matter that much in practice either way.

Avoiding strict adherence to ext4's unicode normalization semantics might pose a problem if userland software were to begin to rely on it, but the likelihood of that is low, since adopting it would mean accepting case insensitivity, which as far as I know, developers on POSIX systems widely abhor.

Since this is a virtual attribute that is not tied to our disk format, if userland software were to begin to exist that explicitly depends on the ext4 unicode normalization semantics, we would have a few options:

* Revert this change at the expense of wine and others like it
* Implement a new ZFS filesystem property and move this behind it to let users enable it on case insensitive filesystems if they want it.
* Implement `casesensitivity=casefold` to convert `ß` to `ss` that also implies normalization=formD, allow `utf8only=off`, switch to exposign this flag when `casesensitivity=casefold` and tell users that want the exact ext4 behavior to create filesystems with `-o casesensitivity=casefold -o normalization=formD -o utf8only=off`. We might also need to investigate whether we would need to update our unicode implementation to utf-12.1 and implement a utf8ver property that must be set at filesystem creation time (since adapting new normalizations from the newer version would cause inaccessible files on existing filesystems).
* Tell them not to write software that way since it is not portable across the vast majority of POSIX filesystems.

I still think it is fine to merge this as is. Whether unicode normalization is turned on should be a user decision and not treating `ß` as `ss` when doing case insensitive lookups is not just more compatible with the behavior userland software expects, but is also something users prefer.

Let me know if anyone wants me to update the commit message to reflect this new information. The commit message is verbose enough as it is that I am hesitant to update it with everything here.

**Comment by ryao** [comment]
### Short version

I did a bit more research on the differences that I spotted. My conclusion that it does not matter for our implementation remains the same.

### Long version

Out of curiosity under the assumption that ext4 and python have the same idea on how to implement casefold, I tried to look at how python implements it to get an idea of the extent to which we differ from ext4. It turns out that it has a helper script generate a database from a file that the unicode.org publishes:

https://github.com/python/cpython/blob/9c197bc8bfa153522927aa03ff854bbc4dce437f/Tools/unicode/makeunicodedata.py#L1175
https://www.unicode.org/Public/14.0.0/ucd/CaseFolding.txt

It uses the C and F lines to determine how it folds. Some quick tests with ext4 suggests that it also follows this document:

Since ext4 uses unicode 12.1, the correct document that describes how ext4's casefold works should be:

https://www.unicode.org/Public/12.1.0/ucd/CaseFolding.txt

```
# touch ß ss SS ẞ
# touch È è
# ls -l
total 0
-rw-r--r-- 1 root root 0 Aug 26 14:36 È
-rw-r--r-- 1 root root 0 Aug 26 14:34 ß
```

The same test on ZFS with casesensitivity=insensitive and normalization=formD gives:

```
# ls -l
total 2
-rw-r--r-- 1 root root 0 Aug 26 14:38 È
-rw-r--r-- 1 root root 0 Aug 26 14:38 ss
-rw-r--r-- 1 root root 0 Aug 26 14:38 ß
-rw-r--r-- 1 root root 0 Aug 26 14:38 ẞ
```

This is just a quick spot test, but the ones that ZFS treated as equivalent were the conversions marked C, while it ignored the ones marked F. If this pattern holds for all of the entries, then our treatment of characters would diverge from ext4 on 104 cases.

However, since we are implementing unicode 5.0 while ext4 is implementing unicode 12.1, it can be assumed that any new C conversions are absent from our codebase, which would give us additional divergence on 476 characters. A quick spot check with `touch ʞ Ʞ` finds this to be correct. There are an additional 43 entries that were added between unicode 12.1 and unicode 14.0. A quick test with `touch 𐖗 𐕰` confirms that ext4's casefold does not affect them.

The collabora blog post suggested that the turkish mappings are also used on ext4, but a quick experimental test with `touch İ i` showed this is not the case, which is what the unicode document suggests.

That said, none of this affects my earlier conclusion that it does not matter as far as implementing this file attribute is concerned.

**Comment by ryao** [comment]
I started to wonder how compatible our case insensitivity with Windows really is, so I asked a friend to do a few quick spot checks where he made files with single character names on Windows 10 21H1 to see when Windows would say the file already exists based on unicode's C definitions. The results are:

* e = E
* È = è
* ß != ẞ
* ʞ != Ʞ
* 𐖗 != 𐕰

This is the same as ZFS. ext4 with the F file attribute set, as previously stated, treats it this way:

* e = E
* È = è
* ß = ẞ
* ʞ = Ʞ
* 𐖗 != 𐕰

**Comment by behlendorf** [comment]
Thanks for the clear and careful analysis of the semantic differences we should expect between this implementation and ext4's behavior.  I've no doubt documenting this will be very helpful in the future!

Personally, my feeling is that it's probably unrealistic to expect identical behavior for the reasons you've mentioned.  We should just document that behavior so it's not a shock.  To that end I'd suggest adding a sentence or two to the `casesensitivity` property description in the man pages.  After that, I'm fine with merging this once you've wrapped up your testing.

**Comment by ryao** [comment]
@behlendorf I have added a note to the man page about the differences and included the pull request number for the brave. I have also amended the commit message to give a basic overview of the additional differences.

I plan to verify a few more aspects of how both MacOS (due to the darling project) and Windows (due to Samba and Wine) implement case insensitive directory lookups on or before Monday, but I expect that to only result in some additional documentation. I will let you know after I have finished doing that.

**Comment by ryao** [comment]
I did testing on a M1 Macbook Air running MacOS 11.4 (it has not yet been updated to 11.6). The testing was done on its default APFS filesystem. The same scripts that worked on Gentoo Linux worked on MacOS, except I had to use curl instead of wget and since touch reported that there were invalid characters, I changed one of the scripts to:

```
(for i in $(cat CaseFolding.txt | grep 'C;' | awk 'BEGIN { FS = ";" } ; {print $1 }'); do  touch $(echo -ne "\U$(printf %08s $i)") || echo $i rejected; done;) > rejected.txt
```

MacOS had some surprises. Since it does unicode normalization, we would expect it to reject invalid character names. However, it goes beyond that by rejecting character names containing unicode characters that have not been defined yet. There were 40 new characters added to the casefold list between Unicode versions 13.0.0 and 14.0.0. It rejected 39 of them, but accepted character 2C2F. 40 characters were not found to have case insensitive matches. 39 of them were of course the 39 it rejected. The 40th was the case insensitive analogy to the character from unicode 14 that it had accepted. Here is the data for those who are interested:

[rejected-mac.txt](https://github.com/openzfs/zfs/files/9457130/rejected-mac.txt)
[results-mac.txt](https://github.com/openzfs/zfs/files/9457131/results-mac.txt)

Note that the erroneous 030C entry is present in results-mac.txt.

A quick comparison shows that there are 3 characters that MacOS treats as case insensitive that ext4 does not. I did a quick spot check with `touch ß ss SS ẞ` and found that MacOS implements casefold. I did not check the other 102 F-marked characters, but I assume that they were all treated the same. The consequence is that ext4's casefold behaves almost identically to MacOS' behavior. We might want to implement casesensitivity=casefold at some point for the MacOS driver. @lundman might also be interested in this information.

@behlendorf While I found some surprises, I do not see anything here that merits blocking this patch. Feel free to merge it at your convenience.

**Comment by ryao** [comment]
By the way, we might want to address the case insensitivity differences between ZFS and Windows at some point, but things have been like this for so long without anyone complaining that I consider it to be a low priority.

If I were to try to address all of these differences, I would:

* Modify our unicode module to support casefold and newer Unicode versions, with an option to reject undefined characters.
* Create a utf8ver property that must be set at filesystem creation.
* Extend utf8only to support a strict value that rejects undefined characters. Being set to strict would allow the utf8ver property to be upgraded to newer versions after filesystem creation.
* Extend casesensitivity to support casefold.

How to handle older drivers mounting filesystems with a utf8ver set would need some thought. I feel like forcing them to mount readonly on unsupported versions would be needed to avoid entries being added to ZAP in ways that newer versions would not allow.

**Comment by ryao** [comment]
@behlendorf On second thought, this is a problem:

> There were 67 characters that ZFS did not treat as case insensitive matches, but Windows did.

Any one of those characters could cause Wine’s directory lookups to fail. ext4’s casefold behavior has the upside that it will never cause a directory lookup to fail when it should not (instead, it can cause directory entry collisions).

This seemed like a great idea when I first thought of it, but we should not make changes that break working userland code on existing filesystems without requiring users to opt-in. Even with that, any software that hits this case would likely cause headaches for the wine project due to users filing bug reports without realizing the root cause.

It might be that wine’s slow path does not treat those characters as insensitive matches since I have not checked, but even if it does, that would be a bug in wine and it does not merit copying that bug into wine’s fast path. Similarly, it could be a good thing that libshare does not inform samba when it is on a case insensitive filesystem.

I am going to withdraw this. We can revisit this if/when someone takes the time to extend ZFS to support a superset of Windows’ case insensitivity.
